### PR TITLE
Feat/epochs processing

### PR DESCRIPTION
**Pull request by vgorkavenko**

### CODE REVIEW

File: src/modules/csm/checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        # copy of
+        # https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#get_block_root_at_slot
+        roots_to_check = []
+        slots = range(
+            self.converter.get_epoch_first_slot(duty_epoch),
+            self.converter.get_epoch_last_slot(EpochNumber(duty_epoch + 1)) + 1
+        )
+        for slot in slots:
+            # TODO: get the magic number from the CL spec
+            if slot + 8192 < self.slot < slot:
```
Comment (madlabman) [inline src/modules/csm/checkpoint.py:123]: To check one more time.

File: src/modules/csm/checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        logger.info({"msg": f"Epoch {duty_epoch} processed in {time.time() - start:.2f} seconds"})
+
+    def _prepare_committees(self, last_finalized_blockstamp: BlockStamp, epoch: int) -> dict:
+        start = time.time()
+        committees = {}
+        for committee in self.cc.get_attestation_committees(last_finalized_blockstamp, EpochNumber(epoch)):
+            validators = []
+            for validator in committee.validators:
+                data = {"index": validator, "included": False}
+                validators.append(data)
```
Comment (madlabman) [inline src/modules/csm/checkpoint.py:160]: Order of insertion is used to track the positions in the committees.

File: src/modules/csm/typings.py [hunk 3]
```diff
@@ -49,14 +51,21 @@ class FramePerformance:
...
-    def dump(self) -> None:
+    def dump(self, epoch: EpochNumber, committees: dict) -> None:
         """Used to persist the current state of the structure."""
+        # TODO: persist the data. no sense to keep it in memory (except of `processed` ?)
+        self.processed.add(epoch)
+        for committee in committees.values():
+            for validator in committee:
+                perf_data = self.aggr_per_val.get(validator['index'], AttestationsAggregate(0, 0))
+                perf_data.assigned += 1
+                perf_data.included += validator['included']
```
Comment (madlabman) [inline src/modules/csm/typings.py:67]: 1 if True else 0

File: src/modules/csm/csm.py [hunk 4]
```diff
@@ -33,19 +36,22 @@ class CSFeeOracle(BaseModule, ConsensusModule):
...
         super().__init__(w3)
-        self.frame_performance: FramePerformance | None
+        self.frame_performance: FramePerformance | None = None
+        # TODO: Feed the cache with the data about the attestations observed so far.
 
     def refresh_contracts(self):
         self.report_contract = self.w3.csm.oracle
 
     def execute_module(self, last_finalized_blockstamp: BlockStamp) -> ModuleExecuteDelay:
         report_blockstamp = self.get_blockstamp_for_report(last_finalized_blockstamp)
```
Comment (madlabman) [inline src/modules/csm/csm.py:46]: Move down so far.

File: src/modules/csm/csm.py [hunk 5]
```diff
@@ -124,25 +130,57 @@ def module_validators_by_node_operators(self, blockstamp: BlockStamp) -> Validat
...
+        start = time.time()
+        for checkpoint in checkpoints:
+            if converter.get_epoch_by_slot(checkpoint.slot) > finalized_epoch:
+                # checkpoint isn't finalized yet, can't be processed
+                break
+            checkpoint.process(last_finalized_blockstamp)
+        delay = time.time() - start
+        logger.info({"msg": f"All epochs processed in {delay:.2f} seconds"})
+
+        self._print_result()
```
Comment (madlabman) [inline src/modules/csm/csm.py:172]: To build report.

File: src/utils/blockstamp.py [hunk 6]
```diff
@@ -25,15 +25,13 @@ def build_blockstamp(
...
-    execution_payload = slot_details.message.body["execution_payload"]
+    execution_payload = slot_details.message.body.execution_payload
 
     return {
         "slot_number": SlotNumber(int(slot_details.message.slot)),
         "state_root": slot_details.message.state_root,
-        "block_number": BlockNumber(int(execution_payload["block_number"])),
-        "block_hash": execution_payload["block_hash"],
-        "block_timestamp": Timestamp(int(execution_payload["timestamp"]))
+        "block_number": BlockNumber(int(execution_payload.block_number)),
```
Comment (madlabman) [inline src/utils/blockstamp.py:34]: Check the need for the conversion.

Comment (vgorkavenko) [reply src/utils/blockstamp.py:34]: needed

File: src/modules/csm/checkpoint.py (hunk 1 above)
Comment (vgorkavenko) [reply src/modules/csm/checkpoint.py]: fixed

File: poetry.lock [hunk 7]
```diff
@@ -1741,41 +1741,18 @@ testutils = ["gitpython (>3)"]
 
 [[package]]
 name = "pyrsistent"
-version = "0.19.3"
+version = "0.16.1"
```
[the remark this answers is not in the document]
Comment (vgorkavenko) [reply poetry.lock]: reverted

**Review by madlabman — APPROVED** [review]

**Pull request merged**